src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,534 @@
+/*
+ * This file follows intgemm and is a means of retrofitting ruy into the intgemm based wiring in
+ * `intgemm_interface.h`. ruy is an inference backend used in tensorflow and android deployment and
+ * has an optimized ARM backend for the multiply operations required. Optimized code for quantize,
+ * unquantize, transpose are added separately to connect the multiply library to marian.
+ */
+
+#pragma once
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include "ruy/platform.h"
+#include "ruy/system_aligned_alloc.h"
+
+#if RUY_PLATFORM_NEON
+#include <arm_neon.h>
+#endif
+
+namespace marian {
+namespace cpu {
+namespace integer {
+
+using Index = unsigned int;
+
+/*
+ * An AlignedVector is similar to intgemm's aligned allocations. Defined here
+ * independently because we are ignoring intgemm path entirely on ARM.
+ */
+template <class T>
+class AlignedVector {
+public:
+  AlignedVector(size_t num_elem)
+      : size_(num_elem),
+        storage_(reinterpret_cast<T *>(ruy::detail::SystemAlignedAlloc(sizeof(T) * num_elem))) {}
+
+  T *begin() { return storage_; }
+  T *data() { return storage_; }
+  size_t size() const { return size_; }
+  size_t memSize() const { return sizeof(T) * size_; }
+
+  // Forbid copy
+  AlignedVector(const AlignedVector &) = delete;
+  AlignedVector &operator=(const AlignedVector &) = delete;
+
+  ~AlignedVector() { ruy::detail::SystemAlignedFree(reinterpret_cast<void *>(storage_)); }
+
+private:
+  size_t size_;
+  T *storage_;
+};
+
+enum class Path { kStandardCpp = 0, kNeon = 1 };
+
+#if RUY_PLATFORM_NEON
+constexpr Path kHighestPath = Path::kNeon;
+#else
+constexpr Path kHighestPath = Path::kStandardCpp;
+#endif
+
+template <enum Path>
+struct Preprocess;
+
+/*
+ * Naive implementation using standard C++ functions. Not optimized using SIMD operations.
+ */
+template <>
+struct Preprocess<Path::kStandardCpp> {
+  static void quantize(const float *input, int8_t *output, float scale, Index rows, Index width) {
+    const Index size = rows * width;
+    for(Index i = 0; i < size; i++) {
+      // Round to nearest after multiplying with scale.
+      float value = roundf(scale * input[i]);
+
+      // Since float can store bigger values, we threshold anything that's gone
+      // higher and can't fit in int8.
+      value = std::max<float>(-127.0f, value);
+      value = std::min<float>(127.0f, value);
+
+      // Finally a static cast.
+      output[i] = static_cast<int8_t>(value);
```

Comment:
Minor: What is the rounding behaviour used here? You might want to be more precise by using `std::round` https://en.cppreference.com/w/cpp/numeric/math/round

Reply:
This isn't minor, it actually impact the quality.  Should be round to nearest.

Reply:
I'm following https://github.com/kpu/intgemm/blob/d3657687c9b84d2ea2ea39b2fac7b89597bde848/test/test.h#L70-L77.  I'm slighly confused at the query.

> float       round ( float arg ); | (1) | (since C++11)(constexpr since C++23)
> -- | -- | --
> float       roundf( float arg ); | (2) | (since C++11)(constexpr since C++23)
> 
> 1-5) Computes the nearest integer value to arg (in floating-point format), rounding halfway cases away from zero, regardless of the current rounding mode.

Reply:
In an attempt to answer: 
(i) round to nearest integer through `roundf` (scale keeps it within range), 
(ii) clamp between [-127, 127]
(iii) cast to `int8_t`.